File: packages/core-dart/test/severity_test.dart
// Severity thresholding (minSeverityLevel) parity suite.
//
// The same table is asserted in core-ts (src/test/severity.test.ts) and
// core-go (routing/severity_test.go) so that all three SDKs filter warnings
// identically: info = 0, warn = 1, error = 2.
import 'package:bluewhale_core/bluewhale_core.dart';
import 'package:test/test.dart';

void main() {
  const baseG = 'GAYCUYT553C5LHVE2XPW5GMEJT4BXGM7AHMJWLAPZP53KJO7EIQADRSI';
  const contractSource =
      'CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC';

  List<String> codes(List<RoutingWarning> warnings) =>
      warnings.map((w) => w.code).toList();

  group('severity weights', () {
    test('uses the normative ordering info=0, warn=1, error=2', () {
      expect(severityOrder, {'info': 0, 'warn': 1, 'error': 2});
      expect(severityWeight(WarningSeverity.info), 0);
      expect(severityWeight(WarningSeverity.warn), 1);
      expect(severityWeight(WarningSeverity.error), 2);
    });

    test('treats unknown or missing severities as info', () {
      expect(severityWeight('fatal'), 0);
      expect(severityWeight(''), 0);
      expect(severityWeight(null), 0);
    });
  });

  group('filterBySeverity', () {
    const mixed = [
      RoutingWarning(
          code: WarningCode.memoIgnoredForMuxed, severity: 'info', message: 'info'),
      RoutingWarning(
          code: WarningCode.nonCanonicalRoutingId, severity: 'warn', message: 'warn'),
      RoutingWarning(
          code: WarningCode.invalidDestination, severity: 'error', message: 'error'),
    ];

    final table = <String, List<String>>{
      'info': [
        'MEMO_IGNORED_FOR_MUXED',
        'NON_CANONICAL_ROUTING_ID',
        'INVALID_DESTINATION',
      ],
      'warn': ['NON_CANONICAL_ROUTING_ID', 'INVALID_DESTINATION'],
      'error': ['INVALID_DESTINATION'],
      'bogus': [
        'MEMO_IGNORED_FOR_MUXED',
        'NON_CANONICAL_ROUTING_ID',
        'INVALID_DESTINATION',
      ],
    };

    table.forEach((min, expected) {
      test('minSeverity=$min keeps $expected', () {
        expect(codes(filterBySeverity(mixed, min)), expected);
      });
    });
  });

  group('extractRoutingSync minSeverityLevel', () {
    final cases = <String, (RoutingInput Function(String?), Map<String, List<String>>)>{
      "G + MEMO_TEXT '007' (warn)": (
        (min) => RoutingInput(
              destination: baseG,
              memoType: 'text',
              memoValue: '007',
              minSeverityLevel: min,
            ),
        {
          'info': ['NON_CANONICAL_ROUTING_ID'],
          'warn': ['NON_CANONICAL_ROUTING_ID'],
          'error': [],
        },
      ),
      'contract source (info)': (
        (min) => RoutingInput(
              destination: baseG,
              memoType: 'id',
              memoValue: '1',
              sourceAccount: contractSource,
              minSeverityLevel: min,
            ),
        {
          'info': ['CONTRACT_SENDER_DETECTED'],
          'warn': [],
          'error': [],
        },
      ),
    };

    cases.forEach((name, spec) {
      final (build, expected) = spec;
      for (final min in ['info', 'warn', 'error']) {
        test('$name @ $min', () {
          expect(codes(extractRoutingSync(build(min)).warnings), expected[min]);
        });
      }
    });

    test('defaults to info when minSeverityLevel is omitted', () {
      final result = extractRoutingSync(
        RoutingInput(destination: baseG, memoType: 'text', memoValue: '007'),
      );
      expect(codes(result.warnings), ['NON_CANONICAL_ROUTING_ID']);
    });
  });
}

File: packages/core-ts/src/test/severity.test.ts
/**
 * Severity thresholding (minSeverityLevel) parity suite.
 *
 * The same table is asserted in core-go (routing/severity_test.go) and
 * core-dart (test/severity_test.dart) so that all three SDKs filter warnings
 * identically: info = 0, warn = 1, error = 2.
 */
import { describe, it, expect } from "vitest";
import {
  extractRouting,
  filterBySeverity,
  severityWeight,
  SEVERITY_ORDER,
} from "../routing/extract";
import type { Warning, WarningSeverity } from "../address/types";

const G_ADDRESS = "GAYCUYT553C5LHVE2XPW5GMEJT4BXGM7AHMJWLAPZP53KJO7EIQADRSI";
const C_SOURCE = "CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC";

const MIXED: Warning[] = [
  { code: "MEMO_IGNORED_FOR_MUXED", severity: "info", message: "info" },
  {
    code: "NON_CANONICAL_ROUTING_ID",
    severity: "warn",
    message: "warn",
    normalization: { original: "007", normalized: "7" },
  },
  {
    code: "INVALID_DESTINATION",
    severity: "error",
    message: "error",
    context: { destinationKind: "C" },
  },
];

const codes = (warnings: { code: string }[]) => warnings.map((w) => w.code);

describe("severity weights", () => {
  it("uses the normative ordering info=0, warn=1, error=2", () => {
    expect(SEVERITY_ORDER).toEqual({ info: 0, warn: 1, error: 2 });
    expect(severityWeight("info")).toBe(0);
    expect(severityWeight("warn")).toBe(1);
    expect(severityWeight("error")).toBe(2);
  });

  it("treats unknown or missing severities as info", () => {
    expect(severityWeight("fatal")).toBe(0);
    expect(severityWeight("")).toBe(0);
    expect(severityWeight(undefined)).toBe(0);
  });
});

describe("filterBySeverity", () => {
  const table: [WarningSeverity, string[]][] = [
    ["info", ["MEMO_IGNORED_FOR_MUXED", "NON_CANONICAL_ROUTING_ID", "INVALID_DESTINATION"]],
    ["warn", ["NON_CANONICAL_ROUTING_ID", "INVALID_DESTINATION"]],
    ["error", ["INVALID_DESTINATION"]],
  ];

  it.each(table)("minSeverity=%s keeps %j", (min, expected) => {
    expect(codes(filterBySeverity(MIXED, min))).toEqual(expected);
  });

  it("treats an unknown threshold as info", () => {
    expect(codes(filterBySeverity(MIXED, "bogus"))).toEqual(codes(MIXED));
  });
});

describe("extractRouting minSeverityLevel", () => {
  const cases: [string, Parameters<typeof extractRouting>[0], Record<WarningSeverity, string[]>][] = [
    [
      "G + MEMO_TEXT '007' (warn)",
      { destination: G_ADDRESS, memoType: "text", memoValue: "007", sourceAccount: null },
      { info: ["NON_CANONICAL_ROUTING_ID"], warn: ["NON_CANONICAL_ROUTING_ID"], error: [] },
    ],
    [
      "contract source (info)",
      { destination: G_ADDRESS, memoType: "id", memoValue: "1", sourceAccount: C_SOURCE },
      { info: ["CONTRACT_SENDER_DETECTED"], warn: [], error: [] },
    ],
  ];

  for (const [name, input, expected] of cases) {
    for (const min of ["info", "warn", "error"] as WarningSeverity[]) {
      it(`${name} @ ${min}`, () => {
        const result = extractRouting({ ...input, minSeverityLevel: min });
        expect(codes(result.warnings)).toEqual(expected[min]);
      });
    }
  }

  it("defaults to info when minSeverityLevel is omitted", () => {
    const result = extractRouting(cases[0][1]);
    expect(codes(result.warnings)).toEqual(["NON_CANONICAL_ROUTING_ID"]);
  });
});
